File: mobile/lib/core/services/socket_service.dart
import 'package:socket_io_client/socket_io_client.dart' as IO;
import 'dart:async';
import '../config/env.dart';
import '../constants/app_constants.dart';
import '../constants/socket_events.dart'; // Phase 4.0

/// Socket Service - Handles Socket.io connections with reconnection logic
class SocketService {
  IO.Socket? _socket;
  bool _isConnected = false;
  bool _isConnecting = false;
  String? _currentToken;
  String? _currentSchoolId;
  int _reconnectAttempts = 0;
  Timer? _reconnectTimer;
  Timer? _heartbeatTimer;

  // Callbacks
  Function(String)? onConnectCallback;
  Function(String)? onDisconnectCallback;
  Function(String, dynamic)? onEventCallback;
  Function(String)? onErrorCallback;

  bool get isConnected => _isConnected;
  bool get isConnecting => _isConnecting;

  /// Connect to Socket.io server
  void connect(String token) {
    if (_isConnecting || (_isConnected && _currentToken == token)) {
      return;
    }

    _currentToken = token;
    _isConnecting = true;
    _reconnectAttempts = 0;

    try {
      _socket?.disconnect();
      _socket?.dispose();

      _socket = IO.io(
        Env.socketUrl,
        IO.OptionBuilder()
            .setTransports(['websocket', 'polling']) // Add polling as fallback
            .setAuth({
              'token': token
            }) // Use auth instead of extraHeaders for Socket.io v4
            .setTimeout(AppConstants.socketTimeout.inMilliseconds)
            .enableReconnection()
            .setReconnectionAttempts(10)
            .setReconnectionDelay(1000)
            .setReconnectionDelayMax(5000)
            .build(),
      );

      _setupEventListeners();
    } catch (e) {
      _isConnecting = false;
      onErrorCallback?.call('Connection error: $e');
    }
  }

  /// Setup default event listeners
  void _setupDefaultListeners() {
    // Default listeners can be added here if needed
  }

  /// Setup event listeners
  void _setupEventListeners() {
    if (_socket == null) return;

    _socket!.onConnect((_) {
      _isConnected = true;
      _isConnecting = false;
      _reconnectAttempts = 0;
      onConnectCallback?.call('Connected to server');

      // Explicitly join room after connection
      if (_currentSchoolId != null) {
        print('🔄 Joining room: school:$_currentSchoolId');
        joinRoom(_currentSchoolId!);
      } else {
        print('⚠️ No school ID available to join room');
      }

      // Listen for room join confirmation
      _socket!.on('JOINED_ROOM', (data) {
        print('✅ Joined room: $data');
      });

      // Start heartbeat
      _startHeartbeat();

      // Setup default event listeners
      _setupDefaultListeners();
    });

    _socket!.onDisconnect((reason) {
      _isConnected = false;
      _isConnecting = false;
      _stopHeartbeat();
      onDisconnectCallback?.call((reason as String?) ?? 'Disconnected');

      // Attempt reconnection with exponential backoff
      if (_currentToken != null) {
        _scheduleReconnect();
      }
    });

    _socket!.onError((error) {
      _isConnecting = false;
      onErrorCallback?.call('Socket error: $error');
    });

    _socket!.onConnectError((error) {
      _isConnecting = false;
      onErrorCallback?.call('Connection error: $error');
      _scheduleReconnect();
    });
  }

  /// Schedule reconnection with exponential backoff
  void _scheduleReconnect() {
    _reconnectTimer?.cancel();

    if (_reconnectAttempts >= AppConstants.maxSocketReconnectAttempts) {
      onErrorCallback?.call('Max reconnection attempts reached');
      return;
    }

    final delay = Duration(
      milliseconds: (AppConstants.connectionRetryDelay.inMilliseconds *
              (1 << _reconnectAttempts))
          .clamp(
        1000,
        30000,
      ),
    );

    _reconnectTimer = Timer(delay, () {
      _reconnectAttempts++;
      if (_currentToken != null) {
        connect(_currentToken!);
      }
    });
  }

  /// Start heartbeat to keep connection alive
  void _startHeartbeat() {
    _stopHeartbeat();
    _heartbeatTimer = Timer.periodic(const Duration(seconds: 30), (_) {
      if (_isConnected) {
        emit(SocketEvents.clientHeartbeat,
            {'timestamp': DateTime.now().toIso8601String()});
      }
    });
  }

  /// Stop heartbeat
  void _stopHeartbeat() {
    _heartbeatTimer?.cancel();
    _heartbeatTimer = null;
  }

  /// Join a room
  void joinRoom(String schoolId) {
    _currentSchoolId = schoolId;
    if (_socket != null && _isConnected) {
      _socket!.emit(SocketEvents.joinRoom, {'schoolId': schoolId});
    }
  }

  /// Listen to an event
  void on(String event, Function(dynamic) handler) {
    if (_socket != null) {
      _socket!.on(event, (data) {
        handler(data);
        onEventCallback?.call(event, data);
      });
    }
  }

  /// Remove event listener
  void off(String event) {
    _socket?.off(event);
  }

  /// Emit an event
  void emit(String event, Map<String, dynamic> data) {
    if (_socket != null && _isConnected) {
      _socket!.emit(event, data);
    }
  }

  /// Disconnect
  void disconnect() {
    _reconnectTimer?.cancel();
    _stopHeartbeat();
    _currentToken = null;
    _currentSchoolId = null;
    _reconnectAttempts = 0;
    _isConnecting = false;
    _isConnected = false;

    if (_socket != null) {
      _socket!.disconnect();
      _socket!.dispose();
      _socket = null;
    }
  }

  /// Dispose all resources
  void dispose() {
    disconnect();
  }
}

File: web/lib/services/socket-service.ts
/**
 * Socket.io client service for real-time events
 */

import { io, Socket } from 'socket.io-client';
import { env } from '../config/env';

export type SocketEvent =
  | 'DRILL_SCHEDULED'
  | 'CRISIS_ALERT'
  | 'DRILL_SUMMARY'
  | 'STUDENT_STATUS_UPDATE'
  | 'ALERT_RESOLVED'
  | 'DRILL_START' // Phase 4.5
  | 'DRILL_END' // Phase 4.5
  | 'DRILL_PARTICIPATION_UPDATE' // Phase 4: Real-time participation updates
  | 'ALERT_CANCEL' // Phase 4.5
  | 'USER_STATUS_UPDATE' // Phase 4.5
  // Phase 3: Student Activity Events
  | 'STUDENT_ACTIVITY_UPDATE' // Real-time student activity updates
  | 'STUDENT_PROGRESS_UPDATE' // Progress changes (XP, badges, score)
  | 'CLASS_ACTIVITY_UPDATE' // Class-wide activity updates
  | 'PARENT_NOTIFICATION' // Notifications for parents
  | 'TEACHER_NOTIFICATION' // Notifications for teachers
  | 'PARENT_VERIFICATION_REQUEST' // Parent verification request
  | 'QR_CODE_SCANNED' // QR code scan notification
  // Phase 201: IoT Device Events
  | 'TELEMETRY_UPDATE' // Real-time telemetry updates from IoT devices
  | 'DEVICE_ALERT'
  | 'SOS_ALERT'
  | 'SOS_SAFE'; // IoT device alerts

export interface SocketEventData {
  [key: string]: any;
}

export interface SocketConnectionStatus {
  connected: boolean;
  connecting: boolean;
  schoolId: string | null;
}

class SocketService {
  private socket: Socket | null = null;
  private listeners: Map<SocketEvent, Set<(data: SocketEventData) => void>> = new Map();
  private schoolId: string | null = null;
  private token: string | null = null;
  private retryCount: number = 0;
  private maxRetries: number = 3;
  private retryTimeout: NodeJS.Timeout | null = null;
  private isConnecting: boolean = false;
  private statusListeners: Set<(status: SocketConnectionStatus) => void> = new Set();

  private notifyStatus() {
    const status = this.getStatus();
    this.statusListeners.forEach((listener) => listener(status));
  }

  getStatus(): SocketConnectionStatus {
    return {
      connected: this.socket?.connected || false,
      connecting: this.isConnecting,
      schoolId: this.schoolId,
    };
  }

  subscribeStatus(listener: (status: SocketConnectionStatus) => void): () => void {
    this.statusListeners.add(listener);
    listener(this.getStatus());
    return () => {
      this.statusListeners.delete(listener);
    };
  }

  connect(schoolId: string, token: string) {
    // Connection Guards: Validate inputs
    if (!schoolId || typeof schoolId !== 'string' || schoolId.trim() === '') {
      console.warn('SocketService: Invalid schoolId provided, skipping connection');
      return;
    }

    if (!token || typeof token !== 'string' || token.trim() === '') {
      console.warn('SocketService: Invalid token provided, skipping connection');
      return;
    }

    // If already connected with same credentials, skip
    if (this.socket?.connected && this.schoolId === schoolId && this.token === token) {
      console.log('SocketService: Already connected with same credentials');
      return;
    }

    // If currently connecting with same credentials, skip
    if (this.isConnecting && this.schoolId === schoolId && this.token === token) {
      console.log('SocketService: Connection already in progress');
      return;
    }

    // Replace transport if credentials changed; keep app-level event subscriptions
    if (this.socket) {
      this.disconnect({ clearListeners: false });
    }

    this.schoolId = schoolId;
    this.token = token;
    this.isConnecting = true;
    this.retryCount = 0;
    this.notifyStatus();

    this.attemptConnection();
  }

  private attemptConnection() {
    if (!this.schoolId || !this.token) {
      this.isConnecting = false;
      return;
    }

    try {
      this.socket = io(env.socketUrl, {
        auth: {
          token: this.token,
        },
        transports: ['websocket', 'polling'],
        reconnection: false, // We handle reconnection manually
        timeout: 5000,
      });

      this.socket.on('connect', () => {
        console.log('Socket connected');
        this.isConnecting = false;
        this.retryCount = 0;
        this.notifyStatus();
        
        // Explicitly join room after connection
        if (this.schoolId) {
          this.joinRoom(this.schoolId);
        }
      });

      // Listen for room join confirmation
      this.socket.on('JOINED_ROOM', (data) => {
        console.log('✅ Joined room:', data);
      });

      this.socket.on('disconnect', (reason) => {
        console.log('Socket disconnected:', reason);
        this.isConnecting = false;
        this.notifyStatus();
        
        // Only retry if it was an unexpected disconnect and we haven't exceeded max retries
        if (reason === 'io server disconnect' || reason === 'transport close') {
          if (this.retryCount < this.maxRetries && this.schoolId && this.token) {
            this.scheduleRetry();
          }
        }
      });

      this.socket.on('connect_error', (error) => {
        console.error('Socket connection error:', error.message);
        this.isConnecting = false;
        this.notifyStatus();
        
        // Retry on connection error
        if (this.retryCount < this.maxRetries && this.schoolId && this.token) {
          this.scheduleRetry();
        } else {
          console.warn('SocketService: Max retries reached, giving up');
        }
      });

      this.socket.on('error', (error) => {
        console.error('Socket error:', error);
      });

      // Register event listeners
      this.setupEventListeners();
    } catch (error) {
      console.error('SocketService: Failed to create socket:', error);
      this.isConnecting = false;
      
      if (this.retryCount < this.maxRetries && this.schoolId && this.token) {
        this.scheduleRetry();
      }
    }
  }

  private scheduleRetry() {
    if (this.retryTimeout) {
      clearTimeout(this.retryTimeout);
    }

    this.retryCount++;
    const delay = 2000 * this.retryCount; // Exponential backoff: 2s, 4s, 6s
    
    console.log(`SocketService: Retrying connection in ${delay}ms (attempt ${this.retryCount}/${this.maxRetries})`);
    
    this.retryTimeout = setTimeout(() => {
      this.retryTimeout = null;
      if (this.schoolId && this.token) {
        this.attemptConnection();
      }
    }, delay);
  }

  private setupEventListeners() {
    if (!this.socket) return;

    const events: SocketEvent[] = [
      'DRILL_SCHEDULED',
      'CRISIS_ALERT',
      'DRILL_SUMMARY',
      'STUDENT_STATUS_UPDATE',
      'ALERT_RESOLVED',
      'DRILL_START', // Phase 4.5
      'DRILL_END', // Phase 4.5
      'DRILL_PARTICIPATION_UPDATE', // Phase 4: Real-time participation updates
      'ALERT_CANCEL', // Phase 4.5
      'USER_STATUS_UPDATE', // Phase 4.5
      // Phase 3: Student Activity Events
      'STUDENT_ACTIVITY_UPDATE',
      'STUDENT_PROGRESS_UPDATE',
      'CLASS_ACTIVITY_UPDATE',
      'PARENT_NOTIFICATION',
      'TEACHER_NOTIFICATION',
      'PARENT_VERIFICATION_REQUEST',
      'QR_CODE_SCANNED',
      // Phase 201: IoT Device Events
      'TELEMETRY_UPDATE',
      'DEVICE_ALERT',
      'SOS_ALERT',
      'SOS_SAFE',
    ];

    events.forEach((event) => {
      this.socket!.on(event, (data: SocketEventData) => {
        const listeners = this.listeners.get(event);
        if (listeners) {
          listeners.forEach((listener) => listener(data));
        }
      });
    });
  }

  private joinRoom(schoolId: string) {
    if (this.socket?.connected) {
      this.socket.emit('JOIN_ROOM', { schoolId, token: this.token });
    }
  }

  on(event: SocketEvent, callback: (data: SocketEventData) => void) {
    if (!this.listeners.has(event)) {
      this.listeners.set(event, new Set());
    }
    this.listeners.get(event)!.add(callback);

    return () => {
      this.off(event, callback);
    };
  }

  off(event: SocketEvent, callback: (data: SocketEventData) => void) {
    const listeners = this.listeners.get(event);
    if (listeners) {
      listeners.delete(callback);
    }
  }

  emit(event: string, data: any) {
    if (this.socket?.connected) {
      this.socket.emit(event, data);
    }
  }

  disconnect(options: { clearListeners?: boolean } = { clearListeners: true }) {
    const clearListeners = options.clearListeners !== false;

    // Safe disconnect - can be called even if not connected
    if (this.retryTimeout) {
      clearTimeout(this.retryTimeout);
      this.retryTimeout = null;
    }

    if (this.socket) {
      try {
        this.socket.removeAllListeners();
        // Only call disconnect() if already connected to avoid "closed before connection established"
        if (this.socket.connected) {
          this.socket.disconnect();
        }
      } catch (error) {
        console.warn('SocketService: Error during disconnect:', error);
      }
      this.socket = null;
    }
    
    this.isConnecting = false;
    this.retryCount = 0;
    if (clearListeners) {
      this.listeners.clear();
    }
    this.schoolId = null;
    this.token = null;
    this.notifyStatus();
  }

  isConnected(): boolean {
    return this.socket?.connected || false;
  }

  isConnectingNow(): boolean {
    return this.isConnecting;
  }
}

export const socketService = new SocketService();

